pixeltable/functions/openai.py

```diff
@@ -1,51 +1,160 @@
 import base64
 import io
-from typing import Optional
+import pathlib
+import uuid
+from typing import Optional, TypeVar, Union, Callable
 
 import PIL.Image
 import numpy as np
+import openai
+import tenacity
+from openai._types import NOT_GIVEN, NotGiven
 
 import pixeltable as pxt
 import pixeltable.type_system as ts
 from pixeltable import env
 from pixeltable.func import Batch
 
 
+def openai_client() -> openai.OpenAI:
+    return env.Env.get().get_client('openai', lambda api_key: openai.OpenAI(api_key=api_key))
+
+
+# Exponential backoff decorator using tenacity.
+# TODO(aaron-siegel): Right now this hardwires random exponential backoff with defaults suggested
+# by OpenAI. Should we investigate making this more customizable in the future?
+def _retry(fn: Callable) -> Callable:
```

Comment:
What's the reason for doing this here? Did you run into some throttling problems?